### DISCUSSION TITLE
Is there a way to implement your own protocol using BEP 10?

### DISCUSSION THREAD

**Discussion by ajnavarro** [anacrolix/torrent#834] [Q&A] [answered]
First of all, thank you for this amazing library.

I was trying to find an easy way to implement my own protocol using BEP 10, maybe implementing a simple interface and injecting it to the client, similar to [webtorrent](https://github.com/webtorrent/bittorrent-protocol#extension-api).

If that is something that is not implemented, I would love to help with it if is something wanted on the project.

Thanks!

**Comment by anacrolix** [comment]
Yes, it's very straightforward. But what is the purpose of your protocol? Are you looking for hooks to add your protocol to a client?

**Reply by ajnavarro** [reply]
Exactly that. I would like to add a new protocol using some kind of hook that I can add to the torrent client, making it able to talk with other peers implementing the same protocol. 

The idea is to create a protocol on top of BEP 10 that will make it possible to negotiate with other peers some data, and after that negotiation, the remote peer will prepare a package with the requested data for the main peer.

**Reply by anacrolix** [reply]
Okay, the kinds of things we'll need will be, a map of extra extension names to offer to remote peer connections, a callback to handle incoming messages for those extensions, and I guess the hardest part, some way to send messages back for those extensions. Do you have something in mind so I can think about how that last one might look?

**Reply by ajnavarro** [reply]
What I was in mind was:
- We discover several peers that might support our custom extension for any kind of method.
- We add these peers to our peer list, marking them as they understand our custom expression. When adding the peer, we can add them with a callback function with some metadata received when the remote answered.
- These peers might give us information about new torrents we want to download.

Maybe we can register a callback when adding a peer at the client level, not associated with any torrent:

```go
	type Extension func(peerID [20]byte, d Data, event string) (Result, error)

	extensions := map[string]Extension{
		"AJ_customext": customext,
	}

	cli, _ := torrent.NewClient(...)
	
	// We discover some new peer using DHT or other sources
	
	cli.AddPeer(torrent.PeerInfo..., extensions)
```

`Extension` can be an interface implementing more useful methods too if needed.

**Reply by anacrolix** [reply]
Yeah I would provide the extension hooks in the ClientConfig to get started. Do you anticipate needing to turn it on just for specific peers, or are you happy to have them be discovered by providing your protocol in their extension handshake? Once you know that a peer supports an extension, what would you need to do with it? Presumably you want to send messages to it using that extension, but what else? What are those messages triggered by? I need to the environment you'll want to operate with when the extension callback is called (like perhaps you will get the PeerConn, with some methods, or some functions you can call to send messages, or maybe another callback to tell you when the PeerConn has closed, etc.)

**Reply by ajnavarro** [reply]
> Do you anticipate needing to turn it on just for specific peers, or are you happy to have them be discovered by providing your protocol in their extension handshake?

I think it is totally ok to have it on for all peers. It is something that is expected from the protocol perspective, and it is also cleaner IMHO.

> Once you know that a peer supports an extension, what would you need to do with it?

In my use case, I'm querying the DHT to find metadata that points to specific peers that might be implementing the custom extension, I don't need to do anything when I connect to a peer that supports my extension by chance, but it might be a good feature to have for other use cases.

> what would you need to do with it? Presumably you want to send messages to it using that extension, but what else? What are those messages triggered by?

The messages will be triggered by an external program, to request data from a specific peer. Example: Send a query with a filter to a specific peer I know has specific torrent files containing CSV files, so after that request, the remote peer can generate a custom result to seed it, and create a response for the custom extension with the generated torrent hash, making possible to the other peer to download the filtered csv content directly.

>  I need to the environment you'll want to operate with when the extension callback is called (like perhaps you will get the PeerConn, with some methods, or some functions you can call to send messages, or maybe another callback to tell you when the PeerConn has closed, etc.)

- PeerConn or some way to identify the remote peer would be needed. 
- The payload of the request will be needed too, to know what to do with that custom extension request. 
- I would need access to the torrent client to start seeding the requested data, but that is something that can depend on the implementation itself.
- Callbacks to know the state of the connection would be great too for handling reconnections or another kind of internal logic.
- The return for the callback can be some kind of response dictionary or error, but it depends on if we want to make the response a sync or async process.

**Reply by anacrolix** [reply]
Asynchronous messages to peers should be easy enough. I think a PeerConn close notification would need to be handled. If you can pin down any specific methods or state you need for a given PeerConn that will help. Looks like everything else fits into the model I had in mind.

**Reply by ajnavarro** [reply]
> If you can pin down any specific methods or state you need for a given PeerConn that will help

I was thinking on:

- OnOpen: when there is a connection with the specific peer
- OnInvalid: when the remote peer doesn't understand the  extension
- OnHandshake: when the remote peer was able to understand the extension
- OnMessage: when the remote peer sends a message
- OnClose: when the connection is closed

Ping me for anything you might need!

**Reply by SemiAccurate** [reply]
> > what would you need to do with it? Presumably you want to send messages to it using that extension, but what else? What are those messages triggered by?
> 
> The messages will be triggered by an external program, to request data from a specific peer. Example: Send a query with a filter to a specific peer I know has specific torrent files containing CSV files, so after that request, the remote peer can generate a custom result to seed it, and create a response for the custom extension with the generated torrent hash, making possible to the other peer to download the filtered csv content directly.

Torrenting is good for when you have multiple peers/clients interested and wanting to share in the same data. But your query with a filter and then the custom result seems to be more of a one to one request-response style of communication. In that light would it not be more suitable to return the result not in the form of a new torrent, but as a custom message inline within the protocol of this extension itself?

So not only would the query be a message within this protocol, but the response is a message with the result data inline back to the requestor.

> > I need to the environment you'll want to operate with when the extension callback is called (like perhaps you will get the PeerConn, with some methods, or some functions you can call to send messages, or maybe another callback to tell you when the PeerConn has closed, etc.)
> 
> * PeerConn or some way to identify the remote peer would be needed.
> * The payload of the request will be needed too, to know what to do with that custom extension request.
> * I would need access to the torrent client to start seeding the requested data, but that is something that can depend on the implementation itself.
> * Callbacks to know the state of the connection would be great too for handling reconnections or another kind of internal logic.
> * The return for the callback can be some kind of response dictionary or error, but it depends on if we want to make the response a sync or async process.

Useful information on how libtorrent implements hooks for dealing with the protocol extension here:
http://libtorrent.org/reference-Plugins.html

And this is libtorrent's writeup on the extension protocol itself:
http://libtorrent.org/extension_protocol.html

**Reply by ajnavarro** [reply]
> Torrenting is good for when you have multiple peers/clients interested and wanting to share in the same data. But your query with a filter and then the custom result seems to be more of a one to one request-response style of communication. In that light would it not be more suitable to return the result not in the form of a new torrent, but as a custom message inline within the protocol of this extension itself?

> So not only would the query be a message within this protocol, but the response is a message with the result data inline back to the requestor.

Giving the result directly instead of the hash to a new torrent is valid, but I think that depends on the use case. Using the same example as I used before, that filter could be cached and the next peer requesting the data will be answered with the same torrent hash and the previous requester will help to seed the result. 

Thanks for the links!

**Reply by anacrolix** [reply]
@SemiAccurate where are the methods to send an extension method in libtorrent?

**Reply by anacrolix** [reply]
This reminds me that https://github.com/anacrolix/torrent/issues/726 is a very important improvement specifically for anacrolix/torrent, which uses eviction a lot.

**Reply by anacrolix** [reply]
> @SemiAccurate where are the methods to send an extension method in libtorrent?

Possibly `peer_connection_handle.send_buffer`?

**Reply by SemiAccurate** [reply]
> > Torrenting is good for when you have multiple peers/clients interested and wanting to share in the same data. But your query with a filter and then the custom result seems to be more of a one to one request-response style of communication. In that light would it not be more suitable to return the result not in the form of a new torrent, but as a custom message inline within the protocol of this extension itself?
> 
> > So not only would the query be a message within this protocol, but the response is a message with the result data inline back to the requestor.
> 
> Giving the result directly instead of the hash to a new torrent is valid, but I think that depends on the use case. Using the same example as I used before, that filter could be cached and the next peer requesting the data will be answered with the same torrent hash and the previous requester will help to seed the result.

But in your example you admit your query yields a "custom result". Therefore it is rather unlikely other peers will be interested in it, as compared to sharing broadly popular torrent contents. Ergo wouldn't it make sense for your use case, which is in the request-response style, to return a response message inline as part of the protocol, rather than generate a new torrent?

> Thanks for the links!

**Reply by SemiAccurate** [reply]
> This reminds me that #726 is a very important improvement specifically for anacrolix/torrent, which uses eviction a lot.

I thought you said that this depends on the storage implementation in use? Which implementation(s) evicts entries a lot? And why do they do so? In general just because of main memory pressure?

Thought even when those storage implementations evict entries from the cache, they're still backed on disk, so why advertise that they're not there when the peer does actually have them?

**Reply by SemiAccurate** [reply]
> @SemiAccurate where are the methods to send an extension method in libtorrent?

@anacrolix here is the complete [code](https://github.com/arvidn/libtorrent/blob/master/src/ut_metadata.cpp) for the ut_metadata extension implementation for an example of how it is done in libtorrent, which includes `send_buffer()`

**Reply by SemiAccurate** [reply]
> > Torrenting is good for when you have multiple peers/clients interested and wanting to share in the same data. But your query with a filter and then the custom result seems to be more of a one to one request-response style of communication. In that light would it not be more suitable to return the result not in the form of a new torrent, but as a custom message inline within the protocol of this extension itself?
> 
> > So not only would the query be a message within this protocol, but the response is a message with the result data inline back to the requestor.
> 
> Giving the result directly instead of the hash to a new torrent is valid, but I think that depends on the use case. Using the same example as I used before, that filter could be cached and the next peer requesting the data will be answered with the same torrent hash and the previous requester will help to seed the result.

@ajnavarro in your example you admit your query yields a "custom result". Therefore it is rather unlikely other peers will be interested in it, as compared to sharing broadly popular torrent contents. Ergo wouldn't it make sense for your use case, which is in the request-response style, to return a response message inline as part of the protocol, rather than generate a new torrent?

**Reply by ajnavarro** [reply]
@SemiAccurate It was just an example of a possible use case. The result can be directly obtained from the request, or if that result has any benefit of being distributed as a torrent, IMHO makes sense to do it to improve decentralization. 

Data can be served in a lot of different ways. That query might be a git request asking for some commits. In that case, the remote peer will have a packfile containing that requested information and many other commits, so the remote can answer by saying that what you want is on this torrent I have. 

Another case is parquet files. You can partition them by a filtered column (the partitioning is done using a folder with names like `gender=female`), so if the query sent by our peer is filtering by that column and maybe many others, we can answer with a torrent file with the content not totally filtered, but narrowed to a specific output that can be distributed to many other peers.

**Reply by SemiAccurate** [reply]
> This reminds me that #726 is a very important improvement specifically for anacrolix/torrent, which uses eviction a lot.

@anacrolix I thought you said that having a cache (and thus evictions) depends on the actual storage implementation in use no? Which anacrolix/torrent implementation(s) evict entries a lot? And why do they do so? In general just because of main memory pressure?

Though even when those storage implementations evict entries from the cache, they're still backed on disk, so why advertise that they're not there when you as a peer do actually have them?

P.S. IIRC Arvid himself has admitted that the BEP 54 lt_donthave extension (which he implemented in libtorrent) is esoteric and not widely used.

P.P.S. I also found this: https://github.com/arvidn/libtorrent/issues/3570#issuecomment-452673405 though I don't know how current it is (ie. if it is now exposed to clients.)

**Reply by anacrolix** [reply]
@SemiAccurate elgatito used to use anacrolix/torrent for the caching and streaming mechanisms it implements, and switched to libtorrent. I suspect in that issue he was curious how to regain the ability to evict data in the same way.

**Reply by SemiAccurate** [reply]
@anacrolix IIRC elgatito used to work on elementum, so your library is rather appropriate for that.

Though when you talk about using eviction a lot, I'm now thinking that you don't maintain a backing store on disk, that it is only an in memory store for that particular (perhaps default) storage implementation of your library?

**Reply by SemiAccurate** [reply]
@anacrolix when you talk about using eviction a lot, I'm now thinking that you don't maintain a backing store on disk, that it is only an in memory store for that particular (perhaps default) storage implementation of your library, is this true?

**Reply by anacrolix** [reply]
In a typical torrent implementation, you only add to the data you have downloaded. You don't complete a torrent and then throw pieces away. anacrolix/torrent supports this, meaning you can operate and download torrents that require more space than you have available. A common example is if you are streaming a large video, you can just download the parts you need while watching it.

**Reply by SemiAccurate** [reply]
> In a typical torrent implementation, you only add to the data you have downloaded. You don't complete a torrent and then throw pieces away. anacrolix/torrent supports this, meaning you can operate and download torrents that require more space than you have available. A common example is if you are streaming a large video, you can just download the parts you need while watching it.

Ah okay! So does that mean that in practice anacrolix/torrent is used mostly for streaming uses cases, unlike other torrent libs?

Is it the only torrent library that does that? Do you know of any others?

Supporting torrenting features unique to streaming is interesting and innovative, though I think you will find that peer-dependent things like BEP 54 lt_donthave aren't useful in practice, because you need others to also support it. If you're the only lib or client that does then you can't really use it or benefit from it. Hmm but also by throwing pieces away a lot it doesn't contribute to the health of the swarm therefore others might conclude it is not a good torrent citizen.

As an aside, I think what you meant by saying that how Reader works is a specific feature of anacrolix/torrent (in the other thread) is that it is unique as a streaming torrenting lib, compared to other torrent libraries, therefore offering specific features like the Reader interface which helps with streaming use cases.

**Reply by anacrolix** [reply]
> Ah okay! So does that mean that in practice anacrolix/torrent is used mostly for streaming uses cases, unlike other torrent libs?

No it works fine for regular use too. It's also used in big crypto projects for whatever kind of prioritization they use, which can't be completely normal.
 
> Is it the only torrent library that does that? Do you know of any others?

I'm pretty sure uTorrent have something similar for their streaming feature, although it's closed source. I think webtorrent have something too, I don't know how sophisticated it is.

> Supporting torrenting features unique to streaming is interesting and innovative, though I think you will find that peer-dependent things like BEP 54 lt_donthave aren't useful in practice, because you need others to also support it. If you're the only lib or client that does then you can't really use it or benefit from it. Hmm but also by throwing pieces away a lot it doesn't contribute to the health of the swarm therefore others might conclude it is not a good torrent citizen.

As an implementation that supports streaming, anacrolix/torrent has gotten this far without needing it. I think you will find it benefits much less than you would expect. If the extension is not available, it only becomes an issue when a peer requests a piece you don't actually have. Disconnecting them, or rejecting their requests are perfectly adequate ways to handle this situation.

> As an aside, I think what you meant by saying that how Reader works is a specific feature of anacrolix/torrent (in the other thread) is that it is unique as a streaming torrenting lib, compared to other torrent libraries, therefore offering specific features like the Reader interface which helps with streaming use cases.

Yes, Reader is an abstraction that makes using the streaming feature very idiomatic in Go. It would still work without streaming but would be awkward and annoying due to lack of prioritization.

**Reply by SemiAccurate** [reply]
> > Ah okay! So does that mean that in practice anacrolix/torrent is used mostly for streaming uses cases, unlike other torrent libs?
> 
> No it works fine for regular use too. It's also used in big crypto projects for whatever kind of prioritization they use, which can't be completely normal.

Big crypto projects? Odd. Which ones were you referring to? Curious, what do they use anacrolix/torrent for?

**Reply by anacrolix** [reply]
@SemiAccurate I think this is offtopic here. I'm not sure it's appropriate for me to discuss the projects.

**Reply by SemiAccurate** [reply]
> @SemiAccurate I think this is offtopic here. I'm not sure it's appropriate for me to discuss the projects.

Oh sorry @anacrolix ! If these big crypto projects using your library are sensitive or not public knowledge then sure I don't care to know their names; I didn't think you worked in crypto.

Though what are the crypto projects using anacrolix/torrent for? Like why would they use your library - or even use torrenting libraries in general - if there's nothing among the features of your library that's special for their use cases. I didn't think blockchains used bittorrent protocol for anything.

(BTW If this is offtopic for this thread, I can transfer our discussion to the [thread](https://github.com/anacrolix/torrent/discussions/844) I recently started, which you just closed, asking for your input. LMK.)

**Comment by anacrolix** [comment, marked as answer]
@ajnavarro user BEP 10 hooks are available as of 8605abc771421206a4cae6bd39f6ca049030d19a. A higher level interface is in discussion elsewhere.
